llama.cpp

```diff
@@ -4036,6 +4045,57 @@ static struct ggml_tensor * llm_build_ffn(
     return cur;
 }
 
+static struct ggml_tensor * llm_build_ffn_mpt_awq(
+        struct ggml_context * ctx,
+         struct ggml_tensor * cur,
+         struct ggml_tensor * up,
+         struct ggml_tensor * up_b,
+         struct ggml_tensor * down,
+         struct ggml_tensor * down_b,
+         struct ggml_tensor * act_scales,
+            llm_ffn_op_type   type_op,
+          llm_ffn_gate_type   type_gate,
+         const llm_build_cb & cb,
+                        int   il) {
+    struct ggml_tensor * tmp = ggml_mul_mat(ctx, up, cur);
+    cb(tmp, "ffn_up", il);
+
+    if (up_b) {
+        tmp = ggml_add(ctx, tmp, up_b);
+        cb(tmp, "ffn_up_b", il);
+    }
+
+    cur = tmp;
+
+    switch (type_op) {
+        case LLM_FFN_GELU_ACT:
+            {
+                cur = ggml_gelu(ctx, cur);
+                cb(cur, "ffn_relu", il);
+                struct ggml_tensor *repeat = ggml_repeat(ctx, act_scales, cur);
+                cb(repeat, "ffn_repeat(scales)", il);
+                cur = ggml_div(ctx, cur, repeat);
```

Comment:
I don't think the `ggml_repeat` is necessary here, `ggml_div` should support full broadcasting.

Reply:
Thank you for your feedback, I updated the code and tested MPT models without any issues.